realm/realm-library/src/objectServer/java/io/realm/PermissionManager.java

```diff
@@ -0,0 +1,517 @@
+    // in order to be able to provide the user with a RealmAsyncTask representation
+    // of the work being done.
+    private RealmAsyncTask addTask(final PermissionManagerAsyncTask task) {
+        if (isReady()) {
+            activateTask(task);
+        } else {
+            delayTask(task);
+            openRealms();
+        }
+
+        return task;
+    }
+
+    // Park the task until all underlying Realms are ready
+    private void delayTask(PermissionManagerAsyncTask task) {
+        delayedTasks.add(task);
+    }
+
+    // Run any tasks that where delayed while the underlying Realms where being opened.
+    // PRECONDITION: Underlying Realms are no longer in the process of being opened.
+    private void runDelayedTasks() {
+        for (PermissionManagerAsyncTask delayedTask : delayedTasks) {
+            activateTask(delayedTask);
+        }
+        delayedTasks.clear();
+    }
+
+    // Activate a task. All tasks are controlled by the Handler in order to make it asynchronous.
+    // PRECONDITION: Underlying Realms are no longer in the process of being opened.
+    private void activateTask(PermissionManagerAsyncTask task) {
+        activeTasks.add(task);
+        handler.post(task);
+    }
+
+    // Open both underlying Realms asynchronously. Once they are both ready, all tasks added in the meantime are
+    // started. If the Realms failed to open correctly any error will be reported through the `Callback.onError` callback.
+    private void openRealms() {
+        if (!openInProgress) {
+            openInProgress = true;
+            managementRealmOpenTask = Realm.getInstanceAsync(managementRealmConfig, new Realm.Callback() {
+                @Override
+                public void onSuccess(Realm realm) {
+                    managementRealm = realm;
+                    checkIfRealmsAreOpened();
+                }
+
+                @Override
+                public void onError(Throwable exception) {
+                    synchronized (errorLock) {
+                        managementRealmError = new ObjectServerError(ErrorCode.UNKNOWN, exception);
+                        checkIfRealmsAreOpened();
+                    }
+                }
+            });
+            permissionRealmOpenTask = Realm.getInstanceAsync(permissionRealmConfig, new Realm.Callback() {
+                @Override
+                public void onSuccess(Realm realm) {
+                    permissionRealm = realm;
+                    checkIfRealmsAreOpened();
+                }
+
+                @Override
+                public void onError(Throwable exception) {
+                    synchronized (errorLock) {
+                        permissionRealmError = new ObjectServerError(ErrorCode.UNKNOWN, exception);
+                        checkIfRealmsAreOpened();
+                    }
+                }
+            });
+        }
+    }
+
+    private void checkIfRealmsAreOpened() {
```

Comment:
maybe rename to something that reflects the `runDelayedTasks` behavior as well. ex `checkIfRealmsAreOpenedAndRunDelayedTasks` 🙃